### ISSUE TITLE
AppSync "subscribe" not working

### ISSUE THREAD

**Issue by mnishiguchi** [ACloudGuru/todoer-demo#1]
@robin-norwood Thanks for your tutorial at [coding-for-cloud-101](https://acloud.guru/course/coding-for-cloud-101/learn/b79bab82-8721-800f-fbe1-93547116d590/69b48a01-9ff6-892e-3b65-bbd18d9da91a/watch). I really enjoyed it and the final demo was very helpful.
I'd like to report an issue I experienced because the demo code did not work as seen in the video.
I googled around and tried to fix it but had no luck as of now.

- Observers are not working; UI won't get updated neither on create or on update.
- `"Variable 'owner' has coerced Null value for NonNull type 'String!'"` error when `subscribe` function is called.

```js
API.graphql(graphqlOperation(subscriptions.onCreateTodo)).subscribe({ ...
```

<img width="900" alt="Screen Shot 2019-10-30 at 8 05 28 AM" src="https://user-images.githubusercontent.com/7563926/67856990-23b80880-faec-11e9-9143-9f81dc360acc.png">

**Comment by zivdamary** [comment]
Yes, @robin-norwood, I too am enjoying studying and learning from you!  I can confirm the same exact issue @mnishiguchi has opened.  I have tried to find where the problem is and have not had success.  Your insight would be much appreciated.  Thanks again for everything!

**Comment by mnishiguchi** [comment]
I spent the whole weekend on fixing this issue. I suspect that this is a bug in the library. I checked a bunch of examples, tried different patterns, looked into generated schema but all looked good to me.

In the end, I decided not to use subscribers because I can do without them. I realized I can do the similar things accessing a promise from graphql operations.

Here is what I ended up with. It is working with no issue.
https://github.com/mnishiguchi/todoer-mnishiguchi/blob/master/src/App.js

I hope this will help anyone.

**Comment by piersas1** [comment]
Hey all,
after looking at this for a bit it seems like the issues is the graphql resource provided in the course is different from the one that is being used in the video .  I was able to get subscriptions to work by passing the owner as part of the graphql query:
   ```
 API.graphql(graphqlOperation(subscriptions.onCreateTodo, { owner: 'xxxxxxx' }))
      .subscribe({...
```

What is being in the video is this: https://github.com/ACloudGuru/todoer-demo/blob/master/src/graphql/subscriptions.js

what is created from the video resources is this: https://gist.github.com/piersas1/8c761e91d19d790836ce7cbf58c10e08

**Comment by alvincrespo** [comment]
@piersas1 You're completely right! This is what I came up with:

```
  const [user, setUser] = useState();
  const [todos, setTodos] = useState([]);
  const [newTodo, setNewTodo] = useState("");

  useEffect(() => {
    const getTodos = async () => {
      const {
        data: {
          listTodos: { items }
        }
      } = await API.graphql(graphqlOperation(queries.listTodos));

      setTodos(items);
    };

    const getUser = async () => {
      const user = await Auth.currentAuthenticatedUser();

      setUser(user);
    };

    getTodos();
    getUser();
  }, []);

  if (user) {
    API.graphql(
      graphqlOperation(subscriptions.onCreateTodo, {
        owner: user.attributes.sub
      })
    ).subscribe({
      next: result => {
        setTodos([...todos, result.value.data.onCreateTodo]);
      }
    });
  }
```

Note: I'm using hooks in my example.

**Comment by astashko** [comment]
It seems like the authenticated user is available in memory, so I did it using `{owner: Auth.user.username}`
